source/Applications/Advanced/TransformViaArucoMarker/TransformViaArucoMarker.cpp

```diff
@@ -0,0 +1,308 @@
+        pointCloud.copyData(
+            reinterpret_cast<Zivid::ColorRGBA *>(rgb.data)); // NOLINT(cppcoreguidelines-pro-type-reinterpret-cast)
+        auto bgr = cv::Mat(pointCloud.height(), pointCloud.width(), CV_8UC4); // NOLINT(hicpp-signed-bitwise)
+        cv::cvtColor(rgb, bgr, cv::COLOR_RGBA2BGR);
+
+        return bgr;
+    }
+
+    std::vector<cv::Point2f> estimate2DMarkerCenters(const aruconano::Marker &markers)
+    {
+        if(markers.empty())
+        {
+            std::cout << "No markers detected";
+            return {};
+        }
+
+        std::vector<cv::Point2f> markerCenters2D;
+        markerCenters2D.reserve(markers.size());
+
+        auto p11 = markers[0];
+        auto p12 = markers[1];
+        auto p21 = markers[3];
+        auto p22 = markers[2];
+
+        const auto backDiagonal{ fitLine(p11, p22) };
+        const auto forwardDiagonal{ fitLine(p21, p12) };
+        const auto xCoord =
+            (forwardDiagonal.intercept - backDiagonal.intercept) / (backDiagonal.slope - forwardDiagonal.slope);
+        const auto yCoord = backDiagonal.slope * xCoord + backDiagonal.intercept;
+        markerCenters2D.emplace_back(cv::Point2f(xCoord, yCoord));
+
+        return markerCenters2D;
+    }
+
+    std::vector<cv::Point2f> get2DMarkerCorners(const aruconano::Marker &markers)
+    {
+        if(markers.empty())
+        {
+            std::cout << "No markers detected";
+            return {};
+        }
+
+        std::vector<cv::Point2f> markerCorners2D;
+        markerCorners2D.reserve(markers.size() * 4);
+
+        markerCorners2D.emplace_back(markers[0]);
+        markerCorners2D.emplace_back(markers[1]);
+        markerCorners2D.emplace_back(markers[2]);
+        markerCorners2D.emplace_back(markers[3]);
+
+        return markerCorners2D;
+    }
+
+    std::vector<cv::Point3d> estimate3DMarkerPoints(const Zivid::PointCloud &pointCloud,
+                                                    const std::vector<cv::Point2f> &markerPoints2D)
+    {
+        if(markerPoints2D.empty())
+        {
+            return {};
+        }
+
+        int w = pointCloud.width();
+        int h = pointCloud.height();
+        const auto points = pointCloud.copyPointsXYZ();
+
+        std::vector<cv::Point3d> markerPoints3D;
+        markerPoints3D.reserve(markerPoints2D.size());
+        for(const auto &point2D : markerPoints2D)
+        {
+            float x = point2D.x;
+            float y = point2D.y;
+
+            int x1 = clamp(static_cast<int>(std::floor(x)), 0, w - 1);
+            int x2 = clamp(static_cast<int>(std::ceil(x)), 0, w - 1);
+            int y1 = clamp(static_cast<int>(std::floor(y)), 0, h - 1);
+            int y2 = clamp(static_cast<int>(std::ceil(y)), 0, h - 1);
+
+            auto q11 = points(y1, x1);
+            auto q12 = points(y2, x1);
+            auto q21 = points(y1, x2);
+            auto q22 = points(y2, x2);
```

Comment:
Why use 'q'? Is it a marker corner? If so, use markerCorner in the var name.

Reply:
This is the same notation as the Wikipedia article explains, https://en.wikipedia.org/wiki/Bilinear_interpolation. I added a comment that this is what I do and therefore I use q, fxy1, fxy2, a, b, c, and d.

Reply:
From wiki: Suppose that we want to find the value of the unknown function f at the point (x, y). It is assumed that we know the value of f at the four points Q11 = (x1, y1), Q12 = (x1, y2), Q21 = (x2, y1), and Q22 = (x2, y2).

Then a comment like 'Getting 2D coordinates for the four points whose 3D coordinates are known' would be helpful perhaps?

Reply:
You mean "Getting 3d coordinates for the four points whose 2D coordinates are known"? x1, y1, etc are the pixel coordinates and hence we get the 3D coordinates with points(x1,y1). Added a sentence explaining this